File: DynamicProgramming/LongestCommonSubsequence.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>  // Needed for reverse()

using namespace std;

// Define the longestCommonSubsequence function before calling it
int longestCommonSubsequence(string text1, string text2) {
    int m = text1.length();
    int n = text2.length();
    
    // Create a dp table initialized with 0s
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

    // Fill the dp table
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (text1[i - 1] == text2[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    // Backtrack to find the LCS string
    int i = m, j = n;
    string res = "";
    while (i > 0 && j > 0) {
        if (text1[i - 1] == text2[j - 1]) {
            res += text1[i - 1];  // Append the matching character
            i--;
            j--;
        } else if (dp[i - 1][j] > dp[i][j - 1]) {
            i--;  // Move upwards
        } else {
            j--;  // Move leftwards
        }
    }

    reverse(res.begin(), res.end());  // Reverse to get the correct order
    cout << "Longest Common Subsequence: " << res << endl;

    return dp[m][n];  // Return the length of LCS
}

int main() {
    string text1 = "abcde";
    string text2 = "ace";
    
    // Call the LCS function and print the length
    int n = longestCommonSubsequence(text1, text2);
    cout << "Length of Longest Common Subsequence: " << n << endl;
    
    return 0;
}

File: DynamicProgramming/Others/LongestCommonSubsequence.java
package DynamicProgramming.Others;

import java.util.*;

public class LongestCommonSubsequence {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        String A = input.nextLine();
        String B = input.nextLine();
        int ans = LCS(A, B);
        System.out.println("LCS Length: " + ans);
        input.close();
    }

    public static int LCS(String A, String B) {
        int m = A.length();
        int n = B.length();
        int[][] dp = new int[m + 1][n + 1];

        // Fill the DP table
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                if (i == 0 || j == 0) {
                    dp[i][j] = 0;
                } else if (A.charAt(i - 1) == B.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }

        // Reconstruct the LCS from the dp table
        int i = m, j = n;
        StringBuilder lCS = new StringBuilder();

        while (i > 0 && j > 0) {
            if (A.charAt(i - 1) == B.charAt(j - 1)) {
                lCS.append(A.charAt(i - 1)); // Add the matching character
                i--;
                j--;
            } else if (dp[i - 1][j] > dp[i][j - 1]) {
                i--; // Move up
            } else {
                j--; // Move left
            }
        }

        // LCS is built in reverse order, so we need to reverse it
        String res = lCS.reverse().toString();
        System.out.println("LCS: " + res);

        return dp[m][n]; // Return the length of the LCS
    }

    // 1143 Medium
    // https://leetcode.com/problems/longest-common-subsequence/description/
    public static int longestCommonSubsequence(String text1, String text2) {
        int m = text1.length(); int n = text2.length();
        int[][] dp = new int[m+1][n+1];
        for (int i = 0; i < m+1; i++) {
            for (int j = 0; j < n+1; j++) {
                if (i==0 || j==0) {
                    dp[i][j] = 0;
                }
                else if (text1.charAt(i-1) == text2.charAt(j-1)) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                }
                else {
                    dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }
        return dp[m][n];
    }
}
